endtoend/endtoend_test.go

```diff
@@ -0,0 +1,524 @@
+	contractAddr   common.Address
+	evaluators     []ev.Evaluator
+}
+
+type beaconNodeInfo struct {
+	processID   int
+	datadir     string
+	rpcPort     uint64
+	monitorPort uint64
+	grpcPort    uint64
+	multiAddr   string
+}
+
+type validatorClientInfo struct {
+	processID   int
+	monitorPort uint64
+}
+
+func runEndToEndTest(t *testing.T, config *end2EndConfig) {
+	tmpPath := path.Join("/tmp/e2e/", uuid.NewRandom().String()[:18])
+	if err := os.MkdirAll(tmpPath, os.ModePerm); err != nil {
+		t.Fatal(err)
+	}
+	config.tmpPath = tmpPath
+	t.Logf("Test Path: %s\n", tmpPath)
+
+	contractAddr, keystorePath, eth1PID := startEth1(t, tmpPath)
+	config.contractAddr = contractAddr
+	beaconNodes := startBeaconNodes(t, config)
+	valClients := initializeValidators(t, config, keystorePath, beaconNodes)
+	processIDs := []int{eth1PID}
+	for _, vv := range valClients {
+		processIDs = append(processIDs, vv.processID)
+	}
+	for _, bb := range beaconNodes {
+		processIDs = append(processIDs, bb.processID)
+	}
+	defer logOutput(t, tmpPath)
+	defer killProcesses(t, processIDs)
+
+	if config.numBeaconNodes > 1 {
+		t.Run("all_peers_connect", func(t *testing.T) {
+			for _, bNode := range beaconNodes {
+				if err := peersConnect(bNode.monitorPort, config.numBeaconNodes-1); err != nil {
+					t.Fatalf("failed to connect to peers: %v", err)
+				}
+			}
+		})
+	}
+
+	beaconLogFile, err := os.Open(path.Join(tmpPath, "beacon-0.log"))
+	if err != nil {
+		t.Fatal(err)
+	}
+	if err := waitForTextInFile(beaconLogFile, "Sending genesis time notification"); err != nil {
+		t.Fatal(err)
+	}
+	conn, err := grpc.Dial("127.0.0.1:4000", grpc.WithInsecure())
+	if err != nil {
+		t.Fatalf("fail to dial: %v", err)
+	}
+	beaconClient := eth.NewBeaconChainClient(conn)
+	nodeClient := eth.NewNodeClient(conn)
+
+	genesis, err := nodeClient.GetGenesis(context.Background(), &ptypes.Empty{})
+	if err != nil {
+		t.Fatal(err)
+	}
+	// Small offset so evaluators perform in the middle of an epoch.
+	epochSeconds := params.BeaconConfig().SecondsPerSlot * params.BeaconConfig().SlotsPerEpoch
+	genesisTime := time.Unix(genesis.GenesisTime.Seconds+int64(epochSeconds/2), 0)
+	currentEpoch := uint64(0)
+	ticker := GetEpochTicker(genesisTime, epochSeconds)
+	for c := range ticker.C() {
+		if c < config.epochsToRun {
+			for _, evaluator := range config.evaluators {
+				// Only run if the policy says so.
+				if !evaluator.Policy(currentEpoch) {
+					continue
+				}
+				t.Run(fmt.Sprintf(evaluator.Name, currentEpoch), func(t *testing.T) {
+					if err := evaluator.Evaluation(beaconClient); err != nil {
+						t.Fatal(err)
+					}
+				})
+			}
+			currentEpoch++
+		} else {
+			ticker.Done()
+			break
+		}
+	}
+
+	if currentEpoch < config.epochsToRun {
+		t.Fatalf("test ended prematurely, only reached epoch %d", currentEpoch)
+	}
+}
+
+// startEth1 starts an eth1 local dev chain and deploys a deposit contract.
+func startEth1(t *testing.T, tmpPath string) (common.Address, string, int) {
```

Comment:
this file is getting quite long, what do you think of separating `eth1` `beacon-node` and `validator` into its own files?

Reply:
Yeah, would make more sense long-term too to split it up.